src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/AuditUtility.cs

```diff
@@ -78,10 +82,99 @@ public AuditUtility(
             }
         }
 
+        private void CheckDownloadDependenciesVulnerabilities(IReadOnlyList<IReadOnlyDictionary<string, IReadOnlyList<PackageVulnerabilityInfo>>> knownVulnerabilities)
+        {
+            Dictionary<PackageIdentity, PackageAuditInfo>? packagesWithKnownVulnerabilities = FindDownloadDependenciesWithKnownVulnerabilities(knownVulnerabilities);
+
+            if (packagesWithKnownVulnerabilities == null)
+            {
+                return;
+            }
+
+            PackageDownloadPackagesWithAdvisory = new(capacity: packagesWithKnownVulnerabilities.Values.Count);
+
+            // no-op checks DGSpec hash, which means the order of everything must be deterministic.
+            foreach ((PackageIdentity package, PackageAuditInfo auditInfo) in packagesWithKnownVulnerabilities.OrderBy(p => p.Key.Id))
+            {
+                PackageDownloadPackagesWithAdvisory.Add(package.Id);
+
+                foreach (var advisory in auditInfo.GraphsPerVulnerability.Keys)
+                {
+                    PackageVulnerabilitySeverity severity = advisory.Severity;
+                    if (severity == PackageVulnerabilitySeverity.Low) { Sev0DirectMatches++; }
+                    else if (severity == PackageVulnerabilitySeverity.Moderate) { Sev1DirectMatches++; }
+                    else if (severity == PackageVulnerabilitySeverity.High) { Sev2DirectMatches++; }
+                    else if (severity == PackageVulnerabilitySeverity.Critical) { Sev3DirectMatches++; }
+                    else { InvalidSevDirectMatches++; }
+                }
+            }
+        }
+
+        private Dictionary<PackageIdentity, PackageAuditInfo>? FindDownloadDependenciesWithKnownVulnerabilities(
+            IReadOnlyList<IReadOnlyDictionary<string, IReadOnlyList<PackageVulnerabilityInfo>>> knownVulnerabilities)
+        {
+            Dictionary<PackageIdentity, PackageAuditInfo>? result = null;
+
+            foreach (var downloadDependency in _downloadDependencies)
+            {
+                List<PackageVulnerabilityInfo>? knownVulnerabilitiesForPackage = GetKnownVulnerabilities(downloadDependency.Name, downloadDependency.VersionRange.MaxVersion!, knownVulnerabilities);
+
+                if (knownVulnerabilitiesForPackage?.Count > 0)
+                {
+                    PackageIdentity packageIdentity = new(downloadDependency.Name, downloadDependency.VersionRange.MinVersion);
```

Comment:
Do we not have this package identity anywhere? 

Can we use the DownloadDependency type? PackageDownload dependency tend to be duplicated among projects, so we should reduce allocations in these codepaths as much as possible.